### ISSUE TITLE
[Phase 4] Evaluation Framework - HEP Metrics

### ISSUE THREAD

**Issue by Sakeeb91** [Sakeeb91/particle-collision-classifier#5]
## Objective

Build comprehensive evaluation framework with HEP-specific metrics and visualization capabilities.

**Part of:** #1

## Dependencies

- #4 (Model Implementation) must be complete

### Metrics Calculator
- [ ] Create `src/evaluation/metrics.py` with MetricsCalculator class
- [ ] Implement standard metrics: AUC, accuracy, precision, recall, F1
- [ ] Add HEP-specific metrics: significance improvement, AMS

### Visualization
- [ ] Create `src/evaluation/visualizer.py` with plotting functions
- [ ] Generate ROC curves with AUC annotation
- [ ] Plot feature importance bar charts
- [ ] Create confusion matrix heatmaps

### Report Generation
- [ ] Create `src/evaluation/reporter.py` for summary reports
- [ ] Generate markdown report with all metrics
- [ ] Save plots as PNG files

### Metrics Calculator

```python
# src/evaluation/metrics.py
import numpy as np
from sklearn.metrics import roc_auc_score, accuracy_score, confusion_matrix

class MetricsCalculator:
    def __init__(self, y_true, y_pred, y_proba=None):
        self.y_true = y_true
        self.y_pred = y_pred
        self.y_proba = y_proba

    def auc(self):
        if self.y_proba is None:
            raise ValueError("Need probabilities for AUC")
        return roc_auc_score(self.y_true, self.y_proba[:, 1])

    def accuracy(self):
        return accuracy_score(self.y_true, self.y_pred)

    def significance_improvement(self, threshold=0.5):
        """
        Calculate significance improvement: s/sqrt(b) ratio.
        Used in HEP to measure discovery potential.
        """
        y_pred_binary = (self.y_proba[:, 1] > threshold).astype(int)
        tp = np.sum((self.y_true == 1) & (y_pred_binary == 1))
        fp = np.sum((self.y_true == 0) & (y_pred_binary == 1))

        if fp == 0:
            return float('inf')
        return tp / np.sqrt(fp)

    def ams_score(self, threshold=0.5, b_reg=10):
        """
        Approximate Median Significance (from Kaggle HIGGS challenge).
        AMS = sqrt(2 * ((s+b+b_reg) * ln(1 + s/(b+b_reg)) - s))
        """
        y_pred_binary = (self.y_proba[:, 1] > threshold).astype(int)
        s = np.sum((self.y_true == 1) & (y_pred_binary == 1))  # true positives
        b = np.sum((self.y_true == 0) & (y_pred_binary == 1))  # false positives

        return np.sqrt(2 * ((s + b + b_reg) * np.log(1 + s / (b + b_reg)) - s))

    def all_metrics(self):
        return {
            'auc': self.auc(),
            'accuracy': self.accuracy(),
            'significance_improvement': self.significance_improvement(),
            'ams': self.ams_score()
        }
```

### Visualizer

```python
# src/evaluation/visualizer.py
import matplotlib.pyplot as plt
from sklearn.metrics import roc_curve, auc

class ResultsVisualizer:
    def __init__(self, save_dir='results/plots'):
        self.save_dir = save_dir

    def plot_roc_curve(self, y_true, y_proba, title='ROC Curve'):
        fpr, tpr, _ = roc_curve(y_true, y_proba[:, 1])
        roc_auc = auc(fpr, tpr)

        plt.figure(figsize=(8, 6))
        plt.plot(fpr, tpr, lw=2, label=f'AUC = {roc_auc:.3f}')
        plt.plot([0, 1], [0, 1], 'k--', lw=1)
        plt.xlim([0.0, 1.0])
        plt.ylim([0.0, 1.05])
        plt.xlabel('False Positive Rate (Background Efficiency)')
        plt.ylabel('True Positive Rate (Signal Efficiency)')
        plt.title(title)
        plt.legend(loc='lower right')
        return plt.gcf()

    def plot_feature_importance(self, model, feature_names, top_n=15):
        importance = model.feature_importances_
        indices = np.argsort(importance)[-top_n:]

        plt.figure(figsize=(10, 6))
        plt.barh(range(len(indices)), importance[indices])
        plt.yticks(range(len(indices)), [feature_names[i] for i in indices])
        plt.xlabel('Feature Importance')
        plt.title('Top Feature Importances')
        return plt.gcf()
```

## Definition of Done

- [ ] All metrics compute correctly (validated against sklearn)
- [ ] ROC curves generate with proper labels
- [ ] Feature importance plots work with XGBoost
- [ ] Markdown reports generate automatically

## Technical Notes

- AMS score uses regularization term b_reg=10 to avoid division issues
- Signal efficiency = TPR, Background rejection = 1 - FPR
- Physics convention: high signal efficiency at low background acceptance
